File: solutions/go/days/1.go
package days

import (
	"fmt"
	"strconv"
	"strings"
)

type Day1 struct{}

func (d Day1) Part1(input []string) string {
	isNumber := func(char string) bool {
		if _, err := strconv.Atoi(char); err == nil {
			return true
		}
		return false
	}

	total := 0
	for _, line := range input {
		firstNumber := ""
		lastNumber := ""
		for _, character := range strings.Split(line, "") {
			if isNumber(character) {
				if firstNumber == "" {
					firstNumber = character
				} else {
					lastNumber = character
				}
			}
		}
		if lastNumber == "" {
			lastNumber = firstNumber
		}
		itemAsNumber, _ := strconv.Atoi(firstNumber + lastNumber)
		total += itemAsNumber
	}

	return fmt.Sprintf("%v", total)
}

func (d Day1) Part2(input []string) string {
	replaceThreeLetterNumbers := func(original []string, i int) (updated []string) {
		updated = original
		threeLetterNumbers := map[string]string{
			"one": "1",
			"two": "2",
			"six": "6",
		}
		searched := original[i-2] + original[i-1] + original[i]
		found := threeLetterNumbers[searched]
		if found != "" {
			updated[i-2] = found
		}
		return updated
	}

	replaceFourLetterNumbers := func(original []string, i int) (updated []string) {
		updated = original
		fourLetterNumbers := map[string]string{
			"four": "4",
			"five": "5",
			"nine": "9",
		}
		searched := original[i-3] + original[i-2] + original[i-1] + original[i]
		found := fourLetterNumbers[searched]
		if found != "" {
			updated[i-3] = found
		}
		return updated
	}

	replaceFiveLetterNumbers := func(original []string, i int) (updated []string) {
		updated = original
		fourLetterNumbers := map[string]string{
			"three": "3",
			"seven": "7",
			"eight": "8",
		}
		searched := original[i-4] + original[i-3] + original[i-2] + original[i-1] + original[i]
		found := fourLetterNumbers[searched]
		if found != "" {
			updated[i-4] = found
		}
		return updated
	}

	var updatedInput []string
	for _, inputItem := range input {
		var replacedChars []string
		for i, char := range strings.Split(inputItem, "") {
			replacedChars = append(replacedChars, char)
			if i > 1 {
				replacedChars = replaceThreeLetterNumbers(replacedChars, i)
			}
			if i > 2 {
				replacedChars = replaceFourLetterNumbers(replacedChars, i)
			}
			if i > 3 {
				replacedChars = replaceFiveLetterNumbers(replacedChars, i)
			}
		}
		updatedInput = append(updatedInput, strings.Join(replacedChars, ""))
	}

	return d.Part1(updatedInput)
}

func NewDay1() Day1 {
	return Day1{}
}

File: solutions/typescript/days/1.ts
import { Run, Day } from "../common";

const Day1: Day = {
  Part1: (input: Array<string>): string => {
    const numbers = ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
    return input
      .reduce((accumulator: number, current: string) => {
        let first = "",
          last = "";
        const chars = current.split("");
        for (let i in chars) {
          const char = chars[i];
          if (numbers.indexOf(char) > -1) {
            if (first === "") {
              first = char;
            } else {
              last = char;
            }
          }
        }
        if (last === "") {
          last = first;
        }
        return accumulator + parseInt(first + last);
      }, 0)
      .toString(10);
  },
  Part2: (input: Array<string>): string => {
    let replace = (
      original: Array<string>,
      i: number,
      map: { [key: string]: string },
      additionalDigit: number
    ): Array<string> => {
      const updated = [...original];
      let search = updated[i - 2] + updated[i - 1] + updated[i];
      if (additionalDigit > 0) search = updated[i - 3] + search;
      if (additionalDigit > 1) search = updated[i - 4] + search;
      const found = map[search];
      if (found !== undefined) updated[i - 2 - additionalDigit] = found;
      return updated;
    };

    const updatedInput: string[] = [];
    for (let i in input) {
      let replaced: string[] = [];
      for (let j in input[i].split("")) {
        replaced.push(input[i][j]);
        if (Number(j) > 1) {
          replaced = replace(
            replaced,
            Number(j),
            {
              one: "1",
              two: "2",
              six: "6",
            },
            0
          );
        }
        if (Number(j) > 2) {
          replaced = replace(
            replaced,
            Number(j),
            {
              four: "4",
              five: "5",
              nine: "9",
            },
            1
          );
        }
        if (Number(j) > 3) {
          replaced = replace(
            replaced,
            Number(j),
            {
              three: "3",
              seven: "7",
              eight: "8",
            },
            2
          );
        }
      }
      updatedInput.push(replaced.join(""));
    }
    return Day1.Part1(updatedInput);
  },
};

Run(Day1);
